lib/Conversion/TosaToSecretArith/TosaToSecretArith.cpp

```diff
@@ -0,0 +1,168 @@
+#include "mlir/include/mlir/IR/ValueRange.h"             // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"              // from @llvm-project
+#include "mlir/include/mlir/Support/LogicalResult.h"     // from @llvm-project
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+
+#define DEBUG_TYPE "tosa-to-secret-arith"
+
+namespace mlir {
+namespace heir {
+namespace tosa {
+
+#define GEN_PASS_DEF_TOSATOSECRETARITH
+#include "lib/Conversion/TosaToSecretArith/TosaToSecretArith.h.inc"
+
+Value createConstantFloat(ImplicitLocOpBuilder &b, double floatValue,
+                          RankedTensorType type) {
+  SmallVector<APFloat> constantValues;
+  constantValues.reserve(type.getNumElements());
+
+  auto elementType = type.getElementType();
+
+  // Create APFloat based on the float type width (this can definitely be made
+  // more elegant somehow)
+  APFloat value(0.0);  // Default initialization
+  if (elementType.isF32()) {
+    value = APFloat(static_cast<float>(
+        floatValue));  // Convert double to float if necessary
+  } else if (elementType.isF64()) {
+    value = APFloat(floatValue);  // Use the double value directly
+  } else {
+    llvm_unreachable("Expected a valid float type for constant creation");
+  }
+
+  // Fill the constant vector with the same value
+  constantValues.assign(type.getNumElements(), value);
+
+  auto constantValuesAttr = DenseElementsAttr::get(type, constantValues);
+  return b.create<arith::ConstantOp>(constantValuesAttr);
+}
+
+struct ConvertTosaSigmoid : public OpRewritePattern<mlir::tosa::SigmoidOp> {
+ private:
+  DataFlowSolver *solver;
+
+ public:
+  ConvertTosaSigmoid(DataFlowSolver *solver, mlir::MLIRContext *context)
+      : OpRewritePattern<mlir::tosa::SigmoidOp>(context), solver(solver) {}
+
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::tosa::SigmoidOp op,
+                                PatternRewriter &rewriter) const override {
+    auto isSecret = [&](Value value) {
+      auto *operandLookup = solver->lookupState<SecretnessLattice>(value);
+      Secretness operandSecretness =
+          operandLookup ? operandLookup->getValue() : Secretness();
+      return (operandSecretness.isInitialized() &&
+              operandSecretness.getSecretness());
+    };
+
+    bool operandIsSecret = isSecret(op.getOperand());
+    if (!operandIsSecret) {
+      return failure();
+    }
+
+    auto inputTensorType =
+        dyn_cast<RankedTensorType>(op.getOperand().getType());
+    if (!inputTensorType) {
+      return failure();
+    }
+
+    auto dimensions = inputTensorType.getShape();
+    auto dataType = inputTensorType.getElementType();
+
+    if (!dyn_cast<FloatType>(dataType)) {
+      return failure();
+    }
+
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
```

Comment:
Maybe a quick comment to note where the approximation came from. Did you get it from a paper? use a tool?

Reply:
+1 - also the bounds it's valid in! one day maybe we can use int range analysis to validate that it applies

Reply:
Sounds good! Done!

Reply:
Maybe it didn't get pushed? I don't see it

Reply:
Got it, paper reference is in the pass doc. thanks!